internal/types/types.go

```diff
@@ -0,0 +1,408 @@
+// simple port of the PQ slice/array type library
+
+package types
```

Comment:
This needs to be at the adapter level with different implementations based on the dialect.